LEAN-3918 LW-FY25-Q2S1: selecting the table, copy/cut and then paste doesn't work

fwaisi opened the pull request.

IslamJMomani commented:
@fwaisi, could you please add a description to the PR to explain what you are doing and why? Thanks!

IslamJMomani reviewed (commented):
If this is about table cut/deletion not being tracked correctly due to applying several delete steps in the same transaction, then this is a general issue that I have encountered in other areas of the editor. I have worked on fixing this in [this PR](https://github.com/Atypon-OpenSource/manuscripts-track-changes-plugin/pull/31).

I noticed that you are checking if the selection is in a table and then skipping remapping based on that, regardless of the operation. This approach doesn’t seem correct to me.

@fwaisi, Could you please review my PR linked above and confirm if it resolves the issue for your case as well?

fwaisi commented:
> If this is about table cut/deletion not being tracked correctly due to applying several delete steps in the same transaction, then this is a general issue that I have encountered in other areas of the editor. I have worked on fixing this in [this PR](https://github.com/Atypon-OpenSource/manuscripts-track-changes-plugin/pull/31).
> 
> I noticed that you are checking if the selection is in a table and then skipping remapping based on that, regardless of the operation. This approach doesn’t seem correct to me.
> 
> @fwaisi, Could you please review my PR linked above and confirm if it resolves the issue for your case as well?

@IslamJMomani, yes your solution fix the issue, but we still need to fix the following condition, because its throw error when try to cut the whole table.
https://github.com/Atypon-OpenSource/manuscripts-track-changes-plugin/blob/72fbf9e7f6b8a04bc8d9e576c5a5ec6705c09643/src/steps/trackTransaction.ts#L95

Will you be able to do it on your pr? so I can close this and resolve the ticket as covered?

fwaisi commented:
@IslamJMomani, @mbartenev-atypon, I have reverted the previous changes, since the changes in the following https://github.com/Atypon-OpenSource/manuscripts-track-changes-plugin/pull/31 will fix the cutting issues. but I have updated the below condition to fail when the uiEvent is cut, because the cutting functionality introduce too many transactions.
https://github.com/Atypon-OpenSource/manuscripts-track-changes-plugin/blob/77e0446dcb3249c9bd20f91c0b6e3263d31d5712/src/steps/trackTransaction.ts#L95

The number of transactions when cutting the whole table:
<img width="1437" alt="Screenshot 2024-09-15 at 4 07 25 PM" src="https://github.com/user-attachments/assets/e648a5da-d73e-4164-8a1f-c2b3e02317e0">

mbartenev-atypon commented on src/steps/trackTransaction.ts:
```diff
@@ -91,7 +91,8 @@ export function trackTransaction(
     // }
-    if (iters > 20) {
+    const uiEvent = tr.getMeta('uiEvent')
+    if (iters > 20 && uiEvent != 'cut') {
```
it would be ideal to still prevent infinite loop somehow even in case of cutting. Even let's say if its a couple of hundred transacations but still a finite number
